include/fiction/technology/sidb_charge_state.hpp

```diff
@@ -122,6 +123,123 @@ charge_configuration_to_string(const std::vector<sidb_charge_state>& charge_dist
 
     return config_str.str();
 }
+/**
+ * The iteration direction over the SiDB states.
+ */
+enum class sidb_state_iter_dir
```

Comment:
Can we maybe avoid the abbreviation here?

Reply:
Do we really need that here? Isn't it possible to get the same effect by using the increment and decrement operator, like in ``bdl_input_iterator`` and ``gray_code_iterator``?

Reply:
I was using it a lot during development of different algorithms in the last months. At the moment it's in use in only a few places actually. Iteration over charge states empowers the generalised thinking over charge states. 

It actually uses regular incr and decr operators like in any iterator and I could move it to `iter` indeed, but honestly the iterator here is so insignificant compares to the ones in there that I feel like it fits here as a little handy utility that comes with `sidb_charge_state` without needing to import another file for a very minor feature.

Maybe I'm misunderstanding your comment actually. I suppose it's not needed indeed, but I use it commonly like this:
```c++
for (const sidb_charge_state cs : sidb_charge_state_iter{})
...
```

and

```c++
for (const sidb_charge_state cs : sidb_charge_state_iter{sidb_charge_state_iter_dir::TO_NEGATIVE})
...
```
 for the other direction.

But I can take the enum out and edit the reverse iteration to non-range based for loops, but then you would need to manually specify POSITIVE as the starting charge state. I actually had two separate iterators before but I thought merging them with a switch was much cleaner

Reply:
I totally understand your point! However, if possible, I would really like to see it in the ``iter`` folder :). I think it is helpful for the structure of ``fiction``, and it might be helpful for other algorithms in the future. Please don't forget unit tests for it. Many thanks!

Reply:
I removed it in the upcoming push. There was very specific use for it for `ClusterExact` that I made earlier, where it was used in the creation of an assignment iterator, where the modularity of an iterator type was useful. However, the constructive approach blows this away completely.

Instead, I just made static constants that `SIDB_CHARGE_STATES_BASE_2` and `SIDB_CHARGE_STATES_BASE_3` over which may be iterated with standard C++ standard library utility. The full-blown iterator character just doesn't fit here.

Note: I made the types `std::vector` instead of `std::array`, since the two static containers differ in length. This is important to ensure the types are compatible, which makes a dynamically evaluated base compatible. (Otherwise you would need to do template arguments that would make things a mess)